### PR TITLE
[WIP] Add BoxList structure in mmdet.

### PR DESCRIPTION
**Pull request by jbwang1997**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add a new BoxList structure in mmdeteciton to support different types of bboxes.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/structures/bbox/base_bbox.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+    _bbox_dim: int = 0
+    SUBMODE_CONVERTERS: dict = None
+
+    def __init__(self,
+                 bboxes: Union[Tensor, np.ndarray],
+                 src_submode: Optional[str] = None,
+                 device: Union[str, torch.device] = None,
+                 dtype: torch.dtype = None) -> None:
+        if isinstance(bboxes, np.ndarray):
+            bboxes = torch.from_numpy(bboxes)
```
Comment (BIGWangYuDong) [inline mmdet/structures/bbox/base_bbox.py:32]: is that possible to add a function to covert bbox from torch to numpy, for example 
```
bbox.numpy()
```

File: mmdet/structures/bbox/base_bbox.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+import numpy as np
+import torch
+from torch import Tensor
+
+
+class BaseBBoxes(metaclass=ABCMeta):
+    """Base class for bounding boxes.
+
+    Args:
+        bboxes(Tensor or np.ndarray): The bbox data.
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_bbox.py:15]: indicate the expected shape.

File: mmdet/structures/bbox/base_bbox.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        src_submode (str, optional): The submode of bboxes. BaseBBoxes will
+            according to src_submode to convert bboxes into regular format.
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_bbox.py:17]: will convert xxx according to xxx.

File: mmdet/structures/bbox/base_bbox.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+import torch
+from torch import Tensor
+
+
+class BaseBBoxes(metaclass=ABCMeta):
+    """Base class for bounding boxes.
+
+    Args:
+        bboxes(Tensor or np.ndarray): The bbox data.
+        src_submode (str, optional): The submode of bboxes. BaseBBoxes will
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_bbox.py:16]: why it is submode rather than mode?

File: mmdet/structures/bbox/base_bbox.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+        cvt_args = [p.name for p in cvt_sig.parameters.values()]
+        assert 'bboxes' in cvt_args and 'direction' in cvt_args, \
+            'Converter must have `bboxes` and `direction` arguments.'
+
+        cls.SUBMODE_CONVERTERS[name] = converter
+
+    @classmethod
+    def register_submode(cls, name, converter=None, force=False):
+        if not isinstance(force, bool):
+            raise TypeError(f'force must be a boolean, but got {type(force)}')
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_bbox.py:67]: seems that they have different return types in different conditions.

File: mmdet/structures/bbox/base_bbox.py (hunk 4 above)
Comment (ZwwWayne) [reply mmdet/structures/bbox/base_bbox.py:16]: We should clarify their differences if necessary.

File: mmdet/structures/bbox/base_bbox.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+from abc import ABCMeta, abstractclassmethod
+from inspect import signature
+from typing import Callable, Optional, Union
+
+import numpy as np
+import torch
+from torch import Tensor
+
+
+class BaseBBoxes(metaclass=ABCMeta):
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_bbox.py:11]: BaseInstanceBox might be a more accurate name.

**Pull request closed without merging**